Applications/FileIO/GocadIO/GocadTSurfaceReader.cpp

```diff
@@ -0,0 +1,525 @@
+            return false;
+        }
+    }
+    return true;
+}
+
+bool GocadTSurfaceReader::readMesh(std::ifstream& in)
+{
+    std::string const mesh_cnt = "-" + std::to_string(_mesh_vec.size() + 1);
+    std::string mesh_name = BaseLib::dropFileExtension(_file_name) + mesh_cnt;
+    if (!parseHeader(in, mesh_name))
+        return false;
+
+    MeshLib::Properties mesh_prop;
+    mesh_prop.createNewPropertyVector<int>(mat_id_name,
+                                           MeshLib::MeshItemType::Cell, 1);
+    std::string line("");
+    while (std::getline(in, line))
+    {
+        std::vector<std::string> str = BaseLib::splitString(line);
+        if (line.empty() || isCommentLine(line))
+            continue;
+        else if (str[0] == "GOCAD_ORIGINAL_COORDINATE_SYSTEM")
+        {
+            Gocad::CoordinateSystem coordinate_system;
+            if (!coordinate_system.parse(in))
+            {
+                ERR("Error parsing coordinate system.");
+                return false;
+            }
+        }
+        else if (str[0] == "GEOLOGICAL_FEATURE" ||
+                 str[0] == "GEOLOGICAL_TYPE" ||
+                 str[0] == "STRATIGRAPHIC_POSITION")
+        {
+            // geological and stratigraphic information - currently ignored
+        }
+        else if (str[0] == "PROPERTY_CLASS_HEADER")
+        {
+            if (!parsePropertyClass(in))
+            {
+                ERR("Error parsing PROPERTY_CLASS_HEADER.");
+                return false;
+            }
+        }
+        else if (str[0] == "PROPERTIES")
+        {
+            if (!parseProperties(in, str, mesh_prop))
+            {
+                ERR("Error parsing PROPERTIES");
+                return false;
+            }
+        }
+        else if (str[0] == "TFACE")
+        {
+            std::vector<MeshLib::Node*> nodes;
+            std::vector<MeshLib::Element*> elems;
+            std::map<std::size_t, std::size_t> node_id_map;
+            INFO ("Parsing surface %s", mesh_name.c_str());
+            if (!parseSurface(in, nodes, elems, node_id_map, mesh_prop))
+            {
+                ERR("Error parsing Surface %s.", mesh_name.c_str());
+                clearData(nodes, elems);
+                return false;
+            }
+            _mesh_vec.push_back(
+                new MeshLib::Mesh(mesh_name, nodes, elems, mesh_prop));
+            return true;
+        }
+        else
+        {
+            WARN("GocadTSurfaceReader::readMesh() - Unknown keyword found: %s",
+                 line.c_str());
+        }
+    }
+    ERR("%s", eof_error.c_str());
+    return false;
+}
+
+MeshLib::Mesh* GocadTSurfaceReader::getData(std::size_t const idx) const
+{
+    if (_mesh_vec.empty())
+    {
+        ERR("Error: No mesh data available.");
+        return nullptr;
+    }
+    if (idx < _mesh_vec.size())
+        return _mesh_vec[idx];
+    ERR("Error: Mesh index (%d) out of range (0, %d).", idx, _mesh_vec.size());
```

Comment:
The max index have to be `_mesh_vec.size() - 1`.

Reply:
:heavy_check_mark: